overlord/patch/patch6_2.go

```diff
@@ -0,0 +1,168 @@
+	Amend            bool `json:"amend,omitempty"`
+	IsAutoRefresh    bool `json:"is-auto-refresh,omitempty"`
+	NoReRefresh      bool `json:"no-rerefresh,omitempty"`
+	RequireTypeBase  bool `json:"require-base-type,omitempty"`
+}
+
+type patch62SnapState struct {
+	SnapType string             `json:"type"`
+	Sequence []*patch62SideInfo `json:"sequence"`
+	Active   bool               `json:"active,omitempty"`
+	Current  snap.Revision      `json:"current"`
+	Channel  string             `json:"channel,omitempty"`
+	patch62Flags
+	Aliases              interface{} `json:"aliases,omitempty"`
+	AutoAliasesDisabled  bool        `json:"auto-aliases-disabled,omitempty"`
+	AliasesPending       bool        `json:"aliases-pending,omitempty"`
+	UserID               int         `json:"user-id,omitempty"`
+	InstanceKey          string      `json:"instance-key,omitempty"`
+	CohortKey            string      `json:"cohort-key,omitempty"`
+	RefreshInhibitedTime *time.Time  `json:"refresh-inhibited-time,omitempty"`
+}
+
+type patch62SnapSetup struct {
+	Channel   string    `json:"channel,omitempty"`
+	UserID    int       `json:"user-id,omitempty"`
+	Base      string    `json:"base,omitempty"`
+	Type      snap.Type `json:"type,omitempty"`
+	PlugsOnly bool      `json:"plugs-only,omitempty"`
+	CohortKey string    `json:"cohort-key,omitempty"`
+	Prereq    []string  `json:"prereq,omitempty"`
+	patch62Flags
+	SnapPath     string           `json:"snap-path,omitempty"`
+	DownloadInfo interface{}      `json:"download-info,omitempty"`
+	SideInfo     *patch62SideInfo `json:"side-info,omitempty"`
+	patch62auxStoreInfo
+	InstanceKey string `json:"instance-key,omitempty"`
+}
+
+type patch62auxStoreInfo struct {
+	Media interface{} `json:"media,omitempty"`
+}
+
+// patch6_2:
+//  - ensure snapd snaps in the snapstate have TypeSnapd for backward compatibility with old snapd snap releases.
+//  - ensure snapd snaps have TypeSnapd in pending install tasks.
+func patch6_2(st *state.State) error {
+	var snaps map[string]*json.RawMessage
+	if err := st.Get("snaps", &snaps); err != nil && err != state.ErrNoState {
+		return fmt.Errorf("internal error: cannot get snaps: %s", err)
+	}
+
+	snapStates := make(map[string]*patch62SnapState)
+	var haveSnapdSnap bool
+
+	// Migrate snapstate; bail out early if we detect snapd snap.
+	// Break up the loop over snaps into two to ease testing.
+	for name, raw := range snaps {
+		var snapst patch62SnapState
+		if err := json.Unmarshal([]byte(*raw), &snapst); err != nil {
+			return err
+		}
+
+		if snapst.SnapType == string(snap.TypeSnapd) {
+			// We can have at most one snapd snap
+			haveSnapdSnap = true
+			break
+		}
+		// cache snap states for the next loop
+		snapStates[name] = &snapst
+	}
+	if !haveSnapdSnap {
+	snapsLoop:
+		for name, snapst := range snapStates {
+			if snapst.SnapType == string(snap.TypeApp) {
+				for _, seq := range snapst.Sequence {
+					if snap.IsSnapd(seq.SnapID) {
+						snapst.SnapType = string(snap.TypeSnapd)
```

Comment:
Will we need an epoch for snapd here if we add this? The `TypeSnapd` was added "Wed Aug 22 2018". AFAICS this means that refreshes from older snapd that get reverted (for whatever reason) will result in a snapd that has an unknown type for the snapd snap - it seems to do not validate the type when we read the snap state so maybe not a problem - just want to make sure we considered it.

Reply:
that would be <=2.35, that's before we released Core 18 though, so a system with such an old snapd shouldn't have a a snapd snap around anyway, no?